### ISSUE TITLE
query.limit(int) causes inconsistent behavior

### ISSUE THREAD

**Issue by dylanharty** [mitsuhiko/flask-sqlalchemy#516]
I am working with a MySQL Database Table data-set that has over 100k rows. flask-sqlalchemy has been causing some issues. The problem is when I add limit() to a query larger than .limit(100) it causes inconsistent behavior in returned results. For example if I .offset(page_number).limit(1000), returned results are random between 1 and < 1000 depending on page(higher or lower page number). 

This code produces stable results(expected results):
`SomeModel.query.offset(offset_number).all()[:count_number]`
this will return result list equal to the count_number (takes a long time to complete)

Respected attempt 1:
`SomeModel.query.offset(offset_number).limit(count_number).all()`
this produces an inconsistent result list being returned. 

Respected attempt 2:
`SomeModel.query.order_by(SomeModel.Time_label.desc()).paginate(page_number, count_number, False).items`
this causes the same inconsistencies as attempt 1 with varying result lists. 

Any suggestions? Am I doing the query wrong or it there a bug in flask-sqlalchemy

**Comment by ThiefMaster** [comment]
If you don't `order_by()` nothing is guaranteed. Since you encounter a problem even without pagination this is a SQLAlchemy issue not specific to Flask-SQLAlchemy.

**Comment by dylanharty** [comment]
attempt 2 has order_by added, still suffers from this issue. So is there a fix?(different syntax) This is terrible for reliable API results.

**Comment by ThiefMaster** [comment]
I'd create a single-file script to reproduce it and report it in the SQLalchemy issue tracker

**Comment by dylanharty** [comment]
I have limited time, so in the mean-time. 
**Steps to recreate issue**
1. Create a valid mysql connection to a table with 30,000+ rows
2. Create a model that inherits from db.model, class SomeModel(db.model)
3. Compare below:
Script to reproduce(py3)
`
page_number = 2 #any number
count_number = 5000 #some big number, less than total rows
offset_number = (page_number * count_number) -  count_number
results = SomeModel.query.offset(offset_number).limit(5001).all()
print(str(results ))
`
this should bring back 5000 results, but it is usually 1 or less than 5000.

**Comment by ThiefMaster** [comment]
Again, look at the SQL that is emitted, check whether it's correct, and whether it returns the expected number of records from your database.

**Comment by dylanharty** [comment]
Ya, I have been debugging that. So
`2017-07-06 13:30:01,930 INFO sqlalchemy.engine.base.Engine (0, 1000)`
this is correct. LIMIT %s, %s will become LIMIT 0, 1000

Then using a len(list) I get 550 on the result set. I have also counted in different ways. 550 count is correct.

**Comment by ThiefMaster** [comment]
very strange. does a `.count()` instead of `.all()` return the same wrong number?

**Comment by dylanharty** [comment]
I wil try, one sec.

**Comment by dylanharty** [comment]
So the .count() returns the correct 1000 count. Maybe an issue in the all() return function?

**Comment by davidism** [comment]
Whatever it is, it's not an issue with anything in Flask-SQLAlchemy.

**Comment by dylanharty** [comment]
@davidism don't want to re-open this issue but I have switched our application from flask-sqlalchemy to the original sqlalchemy. The problem is fixed. So maybe something to look into.

Cheers,

**Comment by eserdk** [comment]
Had the same problem. 
The alternative way to fix it is not to use `.limit()` but to use `.all()[:n]`, where `n` - is desired quantity.

**Comment by Spawn** [comment]
I have the same problem.

**Comment by v4ngelo** [comment]
Not sure if completely related, or even if it corresponds to sqlalchemy or flask-sqlalchemy (I use flask-sqlalchemy), but I had a similar issue with the library performing a 'unique' in my result set while using .all(). This did not happen using the .count() functionality (it always returned the same amount of rows I expected, checking with the same query in the database).

I had to solve it using distinct. So instead of doing:

`self.session.query(self.mapped_class) \
                .order_by(*sql_order if sql_order else default_order)\
                .limit(per_page).offset((page - 1) * per_page)\
                .all()`

had to change it to:

`self.session.query(self.mapped_class).distinct() \
                .order_by(*sql_order if sql_order else default_order)\
                .limit(per_page).offset((page - 1) * per_page)\
                .all()`

And it did the trick.

Just to point to the source, I found that info here -> https://stackoverflow.com/questions/6933478/sqlalchemy-only-one-result-being-returned-when-count-says-there-are-more
